Sources/StreamChat/StateLayer/ChatListState.swift

```diff
@@ -0,0 +1,50 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// Represents a list of channels matching to the specified query.
+@available(iOS 13.0, *)
+public final class ChatListState: ObservableObject {
+    private let observer: Observer
+    private let clientConfig: ChatClientConfig
+    let query: ChannelListQuery
+    
+    init(channels: [ChatChannel], query: ChannelListQuery, clientConfig: ChatClientConfig, database: DatabaseContainer) {
+        self.channels = channels
+        self.clientConfig = clientConfig
+        hasLoadedAllChannels = channels.count < query.pagination.pageSize
+        observer = Observer(query: query, chatClientConfig: clientConfig, database: database)
+        self.query = query
+        
+        observer.start(
+            with: .init(channelsDidChange: { [weak self] changes in
+                guard let self else { return }
+                let channels = await value(forKeyPath: \.channels)
+                await self.setValue(channels.uniquelyApplied(changes, sorting: query.sort), for: \.channels)
+            })
+        )
+    }
+    
+    /// An array of channels for the specified ``ChannelListQuery``.
+    @Published public private(set) var channels = [ChatChannel]()
+    
+    /// True, if all the channels were loaded with paginated loading.
+    @Published public private(set) var hasLoadedAllChannels = false
```

Comment:
How do we distinguish between old channels and new channels?

For example, when jumping to a mid page, we want to know when we finish loading all new channels but also all old channels